packages/remix-react/components.tsx

```diff
@@ -1296,7 +1296,45 @@ export function useMatches() {
  *
  * @see https://remix.run/api/remix#useloaderdata
  */
-export function useLoaderData<T = AppData>(): T {
+
+export type TypedResponse<T> = Response & {
+  json(): Promise<T>;
+};
+
+type DataFunction = (...args: any[]) => unknown; // matches any function
+type DataOrFunction = AppData | DataFunction;
+type JsonPrimitives = string | number | boolean | null;
+type NonJsonPrimitives = undefined | Function | symbol;
+type SerializeType<T> = T extends JsonPrimitives
+  ? T
+  : T extends undefined
+  ? undefined
+  : T extends { toJSON(): infer U }
+  ? U
+  : T extends []
+  ? []
+  : T extends [unknown, ...unknown[]]
+  ? {
+      [k in keyof T]: T[k] extends NonJsonPrimitives
+        ? null
+        : SerializeType<T[k]>;
+    }
+  : T extends (infer U)[]
+  ? SerializeType<U>[]
+  : {
+      [k in keyof T as T[k] extends NonJsonPrimitives
+        ? never
+        : k]: SerializeType<T[k]>;
+    };
```

Comment:
After reading @frontsideair's comment about his `Jsonify` helper function, I thought about [the one from `type-fest`](https://github.com/sindresorhus/type-fest#json).
Couldn't we use that one instead here? 🤔

Reply:
I would advise strongly against using `Jsonify` from `type-fest` for now. I tried it and in edge cases it's less robust than the one in my gist. I opened [an issue ](https://github.com/sindresorhus/type-fest/issues/407) with my findings and will contribute my improvements if they decide to go forward with it.

I haven't tried the one in this PR yet, so I have no comments about it.

Reply:
https://github.com/sindresorhus/type-fest/releases/tag/v2.15.0

Reply:
@colinhacks `type-fest` now has an updated `Jsonify` generic.
https://github.com/sindresorhus/type-fest/releases/tag/v2.15.0

Let's use that one instead of creating our own variant.

Reply:
Agreed. That's sufficiently complex that I'd rather not maintain it 😅

Thanks @sindresorhus!

Reply:
There are still some issues with `Jsonify` that cause failures in the type tests I added to `hook-types-types.tsx`. I've fixed the issues with this PR's implementation discovered by @frontsideair and think we should merge as-is. In parallel I'll create another PR to type-fest fixing some of it's issues; once that gets merged I'll submit a final PR to Remix switching over to typefest's `Jsonify`.

This issues with `Jsonify`:

```ts
Jsonify<unknown>;   
// never
// should be unknown (IMO)

Jsonify<symbol>;   
// never
// should be undefined

Jsonify<Function>;   
// never
// should be undefined

Jsonify<[string, number, Date]>;   
// (string | number)[]
// tuple type not preserved

Jsonify<{ arg1: Function}> 
// {arg: null}, should be {arg: never}
// non-serializable values are omitted from objects
```

The `unknown` problem is a major issue and preserving tuple types is a nice-to-have, everything else is pretty minor. 


As for the issues @frontsideair discovered with `SerializeType`:

> Conversion to primitives for Number, String, Boolean

Fixed

> Turning NonJsonPrimitives to null when in array

Fixed

> Rejecting bigint

Fixed, assuming you by "rejecting" you mean assigning to `never`. In some sense in doesn't matter what the static type is in these scenarios because this will throw a runtime error.

> Omitting entries with symbol keys

Didn't bother with this one yet. I'll fix this in my PR to typefest this is quite minor.
> Rejecting plain NonJsonPrimitives

These should actually be `undefined`. The current version implements this correctly.

Reply:
> > > Rejecting plain NonJsonPrimitives
> 
> > These should actually be undefined. The current version implements this correctly.
> 
> I'm not sure about this. Stringifying any of these returns `undefined` and `JSON.parse` throws, similar to the `bigint` case, so my argument applies here as well. But I'm open to discussion, I'd like to understand why would you like these to be `undefined`?

@frontsideair It's true that an actual `undefined` literal, e.g. inside an array like `[1, 2, undefined]` won't make it back from JSON.stringify and JSON.parse (in that example you'd get `[1, 2, null]`, not a thrown error). But shouldn't simple optional fields in an object be supported? It appears that `undefined` is even being stripped from union types, so an optional field returned from `loader` will come back from `useLoaderData` as if it's always defined, which will very easily lead to a runtime error. For example:

```ts
import { json } from '@remix-run/node';
import { useLoaderData } from '@remix-run/react';

export const loader = () => {
  const user = Math.random() < 0.5 ? { username: 'jo', id: 123 } : undefined;
  return json({ user });
};

export default function Page() {
  const data = useLoaderData<typeof loader>();
  // TypeScript thinks `data.user` is always defined,
  // thus this is a runtime error!
  return <h1>Hello {data.user.username}</h1>;
}

```

This seems like a fairly critical bug in 1.6.5 that can very easily lead to runtime errors.